pil_analyzer/src/evaluator.rs

```diff
@@ -1,104 +1,394 @@
+                        "Expected function but got {e}"
+                    )))?,
+                }
             }
             Expression::MatchExpression(scrutinee, arms) => {
-                let v = self.evaluate(scrutinee);
-                arms.iter()
+                let v = evaluate(scrutinee, locals, symbols)?;
+                let body = arms
+                    .iter()
                     .find_map(|MatchArm { pattern, value }| match pattern {
                         MatchPattern::Pattern(p) => {
-                            (self.evaluate(p) == v).then(|| self.evaluate(value))
+                            // TODO this uses PartialEq. As soon as we have proper match patterns
+                            // instead of value, we can remove the PartialEq requirement on Value.
+                            (evaluate(p, locals, symbols).unwrap() == v).then_some(value)
                         }
-                        MatchPattern::CatchAll => Some(self.evaluate(value)),
+                        MatchPattern::CatchAll => Some(value),
                     })
-                    .expect("No arm matched the value {v}")
+                    .ok_or_else(EvalError::NoMatch)?;
+                evaluate(body, locals, symbols)?
             }
-            Expression::FreeInput(_) => Err("Cannot evaluate free input.".to_string()),
-        }
+            Expression::FreeInput(_) => Err(EvalError::Unsupported(
+                "Cannot evaluate free input.".to_string(),
+            ))?,
+        })
+    }
+
+    fn evaluate_reference<'a, T: FieldElement, C: Custom>(
+        reference: &Reference,
+        locals: &[Rc<Value<'a, T, C>>],
+        symbols: &impl SymbolLookup<'a, T, C>,
+    ) -> Result<Value<'a, T, C>, EvalError> {
+        Ok(match reference {
+            Reference::LocalVar(i, _name) => (*locals[*i as usize]).clone(),
+            Reference::Poly(poly) => symbols.lookup(&poly.name)?,
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use number::GoldilocksField;
+    use pretty_assertions::assert_eq;
+
+    use crate::analyze_string;
+
+    use super::*;
+
+    fn parse_and_evaluate_symbol(input: &str, symbol: &str) -> String {
+        let analyzed = analyze_string::<GoldilocksField>(input);
+        let Some(FunctionValueDefinition::Expression(symbol)) = &analyzed.definitions[symbol].1
+        else {
+            panic!()
+        };
+        evaluate::<_, NoCustom>(symbol, &Definitions(&analyzed.definitions))
+            .unwrap()
+            .to_string()
+    }
+
+    #[test]
+    pub fn arrays_and_strings() {
+        let src = r#"namespace Main(16);
+            let words = ["the", "quick", "brown", "fox"];
+            let translate = |w| match w {
+                "the" => "franz",
+                "quick" => "jagt",
+                "brown" => "mit",
+                "fox" => "dem",
+                _ => "?",
+            };
+            let extract = |arr, i| arr[i];
```

Comment:
This is unused. is there a point in having it in the test?

Reply:
ah, I think I wanted to use it in the next line.